src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+{
+    FabricData fabric(fabric_index);
+    if (CHIP_NO_ERROR == fabric.Load(provider.mStorage))
+    {
+        mNextId = fabric.first_keyset;
+        mCount  = fabric.keyset_count;
+        mIndex  = 0;
+    }
+}
+
+size_t GroupDataProviderImpl::KeysetIteratorImpl::Count()
+{
+    return mCount;
+}
+
+bool GroupDataProviderImpl::KeysetIteratorImpl::Next(Keyset & item)
+{
+    VerifyOrReturnError(mIndex < mCount, false);
+
+    KeysetData keyset(mFabric, mNextId);
+    VerifyOrReturnError(CHIP_NO_ERROR == keyset.Load(mProvider.mStorage), false);
+
+    mIndex++;
+    mNextId            = keyset.next;
+    item.keyset_id     = keyset.keyset_id;
+    item.policy        = keyset.policy;
+    item.num_keys_used = keyset.num_keys_used;
+    memcpy(item.epoch_keys, keyset.epoch_keys, sizeof(item.epoch_keys));
+    return true;
+}
+
+void GroupDataProviderImpl::KeysetIteratorImpl::Release()
+{
+    mProvider.mKeysetIterators.ReleaseObject(this);
+}
+
+//
+// Fabrics
+//
+
+CHIP_ERROR GroupDataProviderImpl::RemoveFabric(chip::FabricIndex fabric_index)
+{
+    FabricData fabric(fabric_index);
+    VerifyOrReturnError(CHIP_NO_ERROR == fabric.Load(mStorage), CHIP_ERROR_KEY_NOT_FOUND);
+
+    // Remove Group Mappings
+
+    EndpointData endpoint(fabric_index, fabric.first_endpoint);
+    size_t endpoint_count = 0;
+
+    do
+    {
+        if (CHIP_NO_ERROR != endpoint.Load(mStorage))
+        {
+            break;
+        }
+        RemoveAllGroupMappings(fabric_index, endpoint.id);
+        endpoint.id = endpoint.next;
+    } while ((++endpoint_count < fabric.endpoint_count) && (kInvalidEndpointId != endpoint.id));
+
+    // Remove States
+
+    StateListData states;
+
+    // Load state list info
+    states.Load(mStorage);
+
+    StateData state(states.first);
+    size_t state_index = 0;
+
+    while (state_index < states.count)
+    {
+        if (CHIP_NO_ERROR != state.Load(mStorage))
+        {
+            break;
+        }
+        if (state.fabric_index == fabric_index)
+        {
+            // Removing the state shifts down the upper states
+            RemoveGroupState(state_index);
+        }
+        else
+        {
+            state_index++;
+        }
+        state.index = state.next;
+    }
+
+    // Remove Keysets
+
+    KeysetData keyset(fabric_index, fabric.first_keyset);
+    size_t keyset_count = 0;
+
+    // Loop until the desired index
+    while (keyset_count < fabric.keyset_count)
+    {
+        if (CHIP_NO_ERROR != keyset.Load(mStorage))
+        {
+            break;
+        }
+        RemoveKeyset(fabric_index, keyset.keyset_id);
+        keyset.keyset_id = keyset.next;
+        keyset_count++;
+    }
+
+    // Remove fabric
+    return fabric.Delete(mStorage);
+}
+
+//
+// General
+//
+
+CHIP_ERROR GroupDataProviderImpl::Decrypt(PacketHeader packetHeader, PayloadHeader & payloadHeader,
+                                          System::PacketBufferHandle && msg)
+{
+    return CHIP_NO_ERROR;
```

Comment:
Please (void) all the unused arguments

Also, since you are receiving a `&& msg`, this will swallow the message and make the r-value reference in outer scope be invalid after the `std::move`. Suggest no taking a `&&`

Reply:
* Unused arguments marked with (void)
* Argument changed as &msg. 
* Return value changed to CHIP_ERROR_NOT_IMPLEMENTED